distribution/xfer/download.go

```diff
@@ -183,15 +183,16 @@ func (ldm *LayerDownloadManager) Download(ctx context.Context, initialRootFS ima
 		progress.Update(progressOutput, descriptor.ID(), "Pulling fs layer")
 
 		var xferFunc DoFunc
-		cacheDir := ldm.cacheDirManager.GetDir(transferKey)
-		cacheDirs[transferKey] = cacheDir
 		if topDownload != nil {
-			xferFunc = ldm.makeDownloadFunc(descriptor, "", topDownload, cacheDir)
+			xferFunc = ldm.makeDownloadFunc(descriptor, "", topDownload, transferKey)
 			defer topDownload.Transfer.Release(watcher)
 		} else {
-			xferFunc = ldm.makeDownloadFunc(descriptor, rootFS.ChainID(), nil, cacheDir)
+			xferFunc = ldm.makeDownloadFunc(descriptor, rootFS.ChainID(), nil, transferKey)
 		}
 		topDownloadUncasted, watcher = ldm.tm.Transfer(transferKey, xferFunc, progressOutput)
+		if watcher.IsOriginal() {
+			cacheDirs[transferKey] = struct{}{}
+		}
```

Comment:
The decision if the dir should be cleaned up or kept is based on whether the overall image download succeeds or fails. The download func doesn't have any knowledge of the overall result of the image download. Even if they did, the release needs to happen after the layers are registered and we know for sure we won't have to re-download them.